source/client/scripts/stars.js

```diff
@@ -1,78 +1,122 @@
-let mouseX = 0;
-let mouseY = 0;
-let stars = [];
-
-const starSize = 3;
-const starSlowness = 700;
-const maxOpacity = 0.8;
-const mouseSpeed = 0.2; // how much does the mouse affect the stars?
-
-const starsContainer = document.getElementById('stars-container');
-
-window.addEventListener('DOMContentLoaded', () => {
-  stars = new Array(400).fill().map(() => ({
-    x: Math.random(),
-    y: Math.random(),
-    z: Math.random(),
-  }));
-
-  for (let i = 0; i < stars.length; i++) {
-    const starElement = document.createElement('div');
-    starElement.className = 'star';
-    updateStar(starElement, i);
-    starsContainer.appendChild(starElement);
-  }
-
-  animate();
-});
+/**
+ * @typedef { {
+ *  xPosPercent: number,
+ *  yPosPercent: number,
+ *  zPosPercent: number,
+ *  htmlElement: HTMLElement
+ * } } Star
+ */
+
+const NUM_STARS = 200;
+const MAX_OPACITY_PERCENT = 0.65;
+const STAR_SPEED_PERCENT = 0.0025;
+const MOUSE_EFFECT_COEFF = 0.025;
+const Z_AXIS_ORIGIN_PX = 250;
 
 /**
- *
+ * @type { {
+ *  xPos: number,
+ *  yPos: number
+ * } }
  */
-function animate() {
-  for (let i = 0; i < stars.length; i++) {
-    stars[i].z = (stars[i].z + 1 / starSlowness) % 1;
-  }
+const mouseState = {
+  xPosPercentFromCenter: 0,
+  yPosPercentFromCenter: 0,
+};
 
-  for (let i = 0; i < stars.length; i++) {
-    updateStar(starsContainer.children[i], i);
-  }
+/**
+ * Updates mouse x and y positions as relative percents of total screen size
+ * in mouseState object, using the center of the screen as the (0, 0) origin
+ * @param { MouseEvent } event mouse movement event passed by listener; passive
+ */
+function updateMousePosition(event) {
+  const { clientWidth, clientHeight } = document.body;
+
+  mouseState.xPosPercentFromCenter = event.clientX / clientWidth - 0.5;
+  mouseState.yPosPercentFromCenter = event.clientY / clientHeight - 0.5;
+} /* updateMousePosition */
+
+/**
+ * Creates and returns new Star objects, which each include initial
+ * x, y, & z positions as well as corresponding HTML elements
+ * @param { number } numStars number of stars to create
+ * @returns { Star[] } array of newly created stars
+ */
+function generateStars(numStars) {
+  return Array.from({ length: numStars }).map(() => {
+    const starEl = document.createElement('div');
+    starEl.className = 'star';
+
+    return {
+      xPosPercent: Math.random(),
+      yPosPercent: Math.random(),
+      zPosPercent: Math.random(),
+      htmlElement: starEl,
+    };
+  });
+} /* generateStars */
 
-  requestAnimationFrame(animate);
-}
+/**
+ * Calculates appropriate current opacity for a star, based off its z position
+ * @param { Star } star target star
+ * @returns { number } percent opacity in range [0, 1]
+ */
+function getPercentOpacity(star) {
+  return star.zPosPercent ** 0.5 * MAX_OPACITY_PERCENT;
```

Comment:
Much more readable, but it changes the fading behavior such that the stars never really fade out, might need to be modified.

Reply:
I see what you mean. My original thinking was that if these are stars, does it really make sense for them to fade out? To me, the fading out makes the stars feel like a somewhat disjoint background while the new implementation makes you go "through" them.

We could perhaps toy around with different speeds and z-axis locations too.

(If anyone else is reading these, please toss in your opinion).

Reply:
I like the no fading in the sense that it seems more realistic if the camera is going "through" the stars. I do agree with Kristin, though, about the reduced amount of stars as it seems somewhat bland.

If we do want it to be somewhat realistic, I would suggest having two types of stars: fast moving stars that quickly spawn in and fall of the edge of the window within a few seconds, and slow moving stars that are much smaller and gradually make their way to the edge. This somewhat gives the perspective of foreground and background, or depth.